doc/neps/nep-XXXX-SIMD-optimizations.rst

```diff
@@ -0,0 +1,249 @@
+
+
+Detailed description
+--------------------
+
+Two new build options are available to ``runtests.py`` and ``setup.py build``.
+The absolute minimum required features to compile are defined by
+``--cpu-baseline``.  For instance, on ``x86_64`` this defaults to ``SSE3``. The
+set of additional intrinsics that can be detected and used as sets of
+requirements to dispatch on are set by ``--cpu-dispatch``. For instance, on
+``x86_64`` this defaults to ``[SSSE3, SSE41, POPCNT, SSE42, AVX, F16C, XOP,
+FMA4, FMA3, AVX2, AVX512F, AVX512CD, AVX512_KNL, AVX512_KNM, AVX512_SKX,
+AVX512_CLX, AVX512_CNL, AVX512_ICL]``. These features are all mapped to a
+c-level boolean array ``npy__cpu_have``, and a c-level convenience function
+``npy_cpu_have(int feature_id)`` queries this array.
+
+The CPU-specific features are then mapped to unversal intrinsics which are
+for all x86 SIMD variants, ARM SIMD variants etc. For example, the
+NumPy universal intrinsic ``npyv_load_u32`` maps to:
+
+*  ``vld1q_u32`` for ARM based NEON
+* ``_mm256_loadu_si256`` for x86 based AVX2 
+* ``_mm512_loadu_si512`` for x86 based AVX-512
+
+Anyone writing a SIMD loop will use the ``npyv_load_u32`` macro instead of the
+architecture specific intrinsic. The code also supplies guard macros for
+compilation and runtime, so that the proper loops can be chosen.
+
+Related Work
+------------
+
+- PIXMAX TBD: what is it?
+- `Eigen`_ is a C++ template library for linear algebra: matrices, vectors,
+  numerical solvers, and related algorithms. It is a higher level-abstraction
+  than the intrinsics discussed here.
+- `xsimd`_ is a header-only C++ library for x86 and ARM that implements the
+  mathematical functions used in the algorithms of ``boost.SIMD``.
+- OpenCV used to have the one-implementation-per-architecture design, but more
+  recently moved to a design that is quite similar to what is proposed in this
+  NEP. The top-level `dispatch code`_ includes a `generic header`_ that is
+  `specialized at compile time`_ by the CMakefile system.
+
+
+Implementation
+--------------
+
+Current PRs:
+
+- `gh-13421 improve runtime detection of CPU features <https://github.com/numpy/numpy/pull/13421>`_
+- `gh-13516: enable multi-platform SIMD compiler optimizations <https://github.com/numpy/numpy/pull/13516>`_
+
+The compile-time and runtime code infrastructure are supplied by the first PR.
+The second adds a demonstration of use of the infrastructure for a loop. Once
+the NEP is approved, more work is needed to write loops using the machnisms
+provided by the NEP.
+
+Alternatives
+------------
+
+A proposed alternative in gh-13516_ is to implement loops for each CPU
+architecture separately by hand, without trying to abstract common patterns in
+the SIMD intrinsics (e.g., have `loops.avx512.c.src`, `loops.avx2.c.src`,
+`loops.sse.c.src`, `loops.vsx.c.src`, `loops.neon.c.src`, etc.). This is more
+similar to what PIXMAX does. There's a lot of duplication here though, and it
```

Comment:
```suggestion
similar to what PIXMAX does. There's a lot of duplication here though, and
```

Reply:
Could it be an an advantage that reusing the same intrinsic could be easier test?

Reply:
Adopting the change. As @r-devulap points out there may be subtle differences between seemingly comparable intrinsics across architectures